File: app/javascript/ErrorReporting.ts
import assertNever from 'assert-never';

export type ErrorReportingLevel = 'fatal' | 'error' | 'warning' | 'log' | 'info' | 'debug';

class ErrorReporting {
  rollbar: ReturnType<(typeof import('rollbar'))['init']> | undefined;
  sentry:
    | {
        instance: import('@sentry/types').Client;
        newScope: () => import('@sentry/types').Scope;
      }
    | undefined;

  static instance: ErrorReporting | undefined;

  static init(...args: ConstructorParameters<typeof ErrorReporting>) {
    ErrorReporting.instance = new ErrorReporting(...args);
  }

  static get() {
    if (ErrorReporting.instance == null) {
      ErrorReporting.init(undefined);
    }

    return ErrorReporting.instance!;
  }

  constructor(currentUserId: string | undefined) {
    if (window.sentryFrontendDSN) {
      import('@sentry/browser').then((Sentry) => {
        const instance = Sentry.init({ dsn: window.sentryFrontendDSN });
        if (instance != null) {
          this.sentry = {
            instance,
            newScope: () => {
              const scope = new Sentry.Scope();
              scope.setUser({ id: currentUserId });
              return scope;
            },
          };
        }
      });
    }

    if (window.rollbarClientAccessToken) {
      import('rollbar').then((Rollbar) => {
        this.rollbar = Rollbar.default.init({
          accessToken: window.rollbarClientAccessToken,
          captureUncaught: true,
          captureUnhandledRejections: true,
          captureIp: 'anonymize',
          payload: {
            environment: import.meta.env.PROD ? 'production' : 'development',
            client: {
              javascript: {
                source_map_enabled: true,
                guess_uncaught_frames: true,
              },
            },
            person: {
              id: currentUserId ?? null,
            },
          },
        });
      });
    }
  }

  report(level: ErrorReportingLevel, error: Error | string, extra?: Record<string, unknown>) {
    const { rollbar, sentry } = this;

    if (rollbar != null) {
      switch (level) {
        case 'fatal':
          rollbar.critical(error, extra);
          break;
        case 'error':
          rollbar.error(error, extra);
          break;
        case 'warning':
          rollbar.warning(error, extra);
          break;
        case 'log':
          rollbar.info(error, extra);
          break;
        case 'info':
          rollbar.info(error, extra);
          break;
        case 'debug':
          rollbar.debug(error, extra);
          break;
        default:
          assertNever(level, true);
          rollbar?.error(error, extra);
      }
    }

    if (sentry != null) {
      const { tags, ...other } = extra ?? {};
      const scope = sentry.newScope();
      scope.setContext('extra', other);
      scope.setTags(tags as Record<string, string>);

      if (error instanceof Error) {
        sentry.instance.captureException(error, undefined, scope);
      } else {
        sentry.instance.captureMessage(error, level, undefined, scope);
      }
    }
  }

  fatal(error: Error | string, extra?: Record<string, unknown>) {
    this.report('fatal', error, extra);
  }

  error(error: Error | string, extra?: Record<string, unknown>) {
    this.report('error', error, extra);
  }

  warning(error: Error | string, extra?: Record<string, unknown>) {
    this.report('warning', error, extra);
  }

  log(error: Error | string, extra?: Record<string, unknown>) {
    this.report('log', error, extra);
  }

  info(error: Error | string, extra?: Record<string, unknown>) {
    this.report('info', error, extra);
  }

  debug(error: Error | string, extra?: Record<string, unknown>) {
    this.report('debug', error, extra);
  }
}

export const initErrorReporting = ErrorReporting.init;
const errorReporting = ErrorReporting.get;

export default errorReporting;

File: app/services/error_reporting.rb
class ErrorReporting
  RUBY_LOGGER_LEVELS = {
    fatal: Logger::FATAL,
    error: Logger::ERROR,
    warning: Logger::WARN,
    log: Logger::INFO,
    info: Logger::INFO,
    debug: Logger::DEBUG
  }

  ROLLBAR_LEVELS = {
    fatal: 'critical',
    error: 'error',
    warning: 'warning',
    log: 'info',
    info: 'info',
    debug: 'debug'
  }

  SENTRY_LEVELS = {
    fatal: :fatal,
    error: :error,
    warning: :warning,
    log: :log,
    info: :info,
    debug: :debug
  }

  def self.report(level, error, **extra)
    Rails.logger.add(RUBY_LOGGER_LEVELS.fetch(level), error:, **extra)

    if Rollbar.configuration.enabled
      Rollbar.log(ROLLBAR_LEVELS.fetch(level), error, **extra)
    end

    return unless Sentry.initialized?
    send_to_sentry(error, level: SENTRY_LEVELS.fetch(level), **extra)
  end

  %i[fatal error warning log info debug].each do |level|
    define_singleton_method level do |error, **extra|
      report(level, error, **extra)
    end
  end

  def self.warn(error, **extra)
    report(:warning, error, **extra)
  end

  def self.critical(error, **extra)
    report(:fatal, error, **extra)
  end

  def self.send_to_sentry(error, level:, tags: nil, **extra)
    Sentry.with_scope do |scope|
      scope.set_context("extra", extra) if extra.present?
      scope.set_tags(tags) if tags.present?

      case error
      when Exception
        Sentry.capture_exception(error, level:)
      else
        Sentry.capture_message(error, level:)
      end
    end
  end
end
